sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1690,11 +1690,10 @@ object PushPredicateThroughNonJoin extends Rule[LogicalPlan] with PredicateHelpe
    */
   private def canPushThroughCondition(plan: LogicalPlan, condition: Expression): Boolean = {
     val attributes = plan.outputSet
-    val matched = condition.find {
-      case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
-      case _ => false
+    condition.exists {
```

Comment:
this one looks a bit confusing. Does it really match the previous logic?

Reply:
This is how I derived this, the original logic same as:

```scala
condition.find {
  case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
  case _ => false
}.isEmpty
```
It is equivalent to

```scala
!condition.find {
  case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
  case _ => false
}.noEmpty
```

-->

```scala
!condition.exists {
  case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
  case _ => false
}
```

->
```
condition.exists {
  case s: SubqueryExpression => !s.plan.outputSet.intersect(attributes).nonEmpty
  case _ => !false
}
```

->
```
condition.exists {
  case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).isEmpty
  case _ => true
}
```

Reply:
hmm... there is something wrong with this derivation process

Reply:
I revert this to:

```scala
!condition.exists {
      case s: SubqueryExpression => s.plan.outputSet.intersect(attributes).nonEmpty
      case _ => false
    }
```
@cloud-fan